web/frontend/src/components/keyboard-input.vue

```diff
@@ -1,108 +1,108 @@
+  return ({
+    w: 'w',
+    a: 'a',
+    s: 's',
+    d: 'd',
+    arrowup: 'w',
+    arrowleft: 'a',
+    arrowdown: 's',
+    arrowright: 'd',
+  } as Record<string, Keys>)[key.toLowerCase()] ?? null;
 };
 
-const keyIcons = {
-  forward: mdiArrowUp,
-  left: mdiRestore,
-  backward: mdiArrowDown,
-  right: mdiReload,
+const emitKeyDown = (key: Keys) => {
+  if (!isActive) {
+    // eslint-disable-next-line no-use-before-define
+    toggleKeyboard(true);
+  }
+  pressedKeys[key] = true;
+  emit('keydown', key);
 };
 
-const keysLayout = [['forward'], ['left', 'backward', 'right']] as const;
-
-const setKeyPressed = (key: 'forward' | 'left' | 'backward' | 'right', value = true) => {
-  pressedKeys[key] = value;
-  emit('keyboard-ctl', pressedKeys);
+const emitKeyUp = (key: Keys) => {
+  pressedKeys[key] = false;
+  emit('keyup', key);
 };
 
-const onUseKeyboardNav = (event: KeyboardEvent) => {
+const handleKeyDown = (event: KeyboardEvent) => {
   event.preventDefault();
+  event.stopPropagation();
+
+  const key = normalizeKey(event.key);
 
-  const down = event.type === 'keydown';
-
-  switch (event.key.toLowerCase()) {
-    case 'arrowleft':
-    case 'a': {
-      return setKeyPressed('left', down);
-    }
-    case 'arrowright':
-    case 'd': {
-      return setKeyPressed('right', down);
-    }
-    case 'arrowup':
-    case 'w': {
-      return setKeyPressed('forward', down);
-    }
-    case 'arrowdown':
-    case 's': {
-      return setKeyPressed('backward', down);
-    }
+  if (key === null || pressedKeys[key]) {
+    return;
   }
-};
 
-const addKeyboardListeners = () => {
-  isActive = true;
-  window.addEventListener('keydown', onUseKeyboardNav, false);
-  window.addEventListener('keyup', onUseKeyboardNav, false);
+  emitKeyDown(key);
 };
 
-const removeKeyboardListeners = () => {
-  isActive = false;
-  window.removeEventListener('keydown', onUseKeyboardNav);
-  window.removeEventListener('keyup', onUseKeyboardNav);
-};
+const handleKeyUp = (event: KeyboardEvent) => {
+  event.preventDefault();
+  event.stopPropagation();
 
-const stopBase = () => {
-  const req = new baseApi.StopRequest();
-  req.setName(props.name);
-  window.baseService.stop(req, new grpc.Metadata(), displayError);
+  const key = normalizeKey(event.key);
+  if (key !== null) {
+    emitKeyUp(key);
+  }
 };
 
-const toggleKeyboard = () => {
-  if (isActive) {
-    removeKeyboardListeners();
-    stopBase();
+const toggleKeyboard = (nowActive: boolean) => {
+  if (nowActive) {
+    window.addEventListener('keydown', handleKeyDown, false);
+    window.addEventListener('keyup', handleKeyUp, false);
   } else {
-    addKeyboardListeners();
+    window.removeEventListener('keydown', handleKeyDown);
+    window.removeEventListener('keyup', handleKeyUp);
   }
+
+  isActive = nowActive;
+};
+
+const handlePointerDown = (key: Keys) => {
+  emitKeyDown(key);
+  window.addEventListener('pointerup', () => emitKeyUp(key), { once: true });
```

Comment:
This specifically fixes one of the issues people are encountering: dragging a mouse away from a button causes the robot to drive forever, off into the sunset, never to be seen again. 

If we instead apply the keyup event to the window, the mouse can be anywhere and the input will still be canceled.